docs/modules/ROOT/pages/adr/0034-vshnredis-high-availability.adoc

```diff
@@ -0,0 +1,100 @@
+= ADR 0034 - Redis High Availability /  Clustering
+:adr_author:    Nicolas Bigler
+:adr_owner:     Schedar
+:adr_reviewers: 
+:adr_date:      2025-06-04
+:adr_upd_date:  2025-06-04
+:adr_status:    draft
+:adr_tags:      redis,ha
+
+include::partial$adr-meta.adoc[]
+
+[NOTE]
+.Summary
+====
+This ADR evaluates different Redis HA deployment options on Kubernetes to support a production-grade, fault-tolerant Redis setup. The goal is to ensure high availability, resilience to node failures, and ease of management/operations.
+====
+
+== Context
+
+Requirements to the Redis HA architecture:
+
+* Minimizes downtime during failover
+* Supports Kubernetes-native patterns
+* Is operationally manageable
+* Is compatible with our existing infrastructure and client usage patterns
+
+=== Considered Options
+
+Option 1:  Redis Sentinel + Redis Master/Replica::
+
+image::adr-0034-redis-sentinel.png[Redis Master/Replica & Sentinel Example]
+
+* Description: Traditional Redis HA using Sentinel to monitor and promote replicas when the master fails.
+* Pros:
+  ** Well-established model
+  ** Supported by Redis directly
+  ** Easy to deploy and manage with the Bitnami Helm chart
+  ** Supports Kubernetes-native patterns
+* Cons:
+  ** Requires client that supports sentinel
+  ** Can't be used when exposing Redis outside of Kubernetes (for example Loadbalancer service) and therefore can't be used by Servala.
+
+Option 2: Redis Cluster Mode::
+
+* Description: Redis Cluster offers built-in sharding and high availability with multiple masters and replicas.
+* Pros:
+  ** Native Redis HA and scalability
+  ** No need for Sentinel
+* Cons:
+  ** Requires client-side cluster awareness
+  ** Higher operational complexity: need to manage multiple masters and replicas
+
+Option 3: Redis Master/Replica with TCP Proxy in Front (for example HAProxy, Envoy)::
+
+image::adr-0034-redis-sentinel-haproxy.png[Redis Master/Replica & Sentinel & HAProxy Example]
+
+* Description: Deploy Redis in master-replica mode, expose the Redis master behind a TCP load balancer or proxy that can detect failure and route to the promoted master.
+* Pros:
+  ** No need for Sentinel
+  ** Client remains unaware of master failover; proxy handles redirection
+  ** Greater flexibility in failover and health checks
+  ** Can be exposed outside of Kubernetes (for example Loadbalancer service) and therefore be used by Servala
+* Cons:
+  ** Requires custom failover scripts or integration with Sentinel to update routes
+  ** Slight increase in latency due to proxy hop
+  ** Operational complexity: proxy config must track Redis role changes
+
+Option 4: Redis Master/Replica with Label-based Service Selector::
```

Comment:
Did you do a test to see how long K8s takes to actually change the traffic to another pod?

For patroni it takes about 10-15 seconds. Patroni does it similarly, but there's also a lot of PostgreSQL failover shenanigans going on. So it might even be faster for Redis, but probably slower than using a proper proxy.

Reply:
It's actually pretty fast. Here is a screencast of this in action:

https://github.com/user-attachments/assets/9301abc7-5b73-4a47-b7c3-6001993fd6a1